src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1277,8 +1290,9 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 // pipes.   
                 Debug.Assert(_readPos == 0 && _readLen == 0, "Win32FileStream must not have buffered data here!  Pipes should be unidirectional.");
 
-                if (_writePos > 0)
-                    FlushWrite(false);
+                if(_writePos > 0)
+                    return FlushWriteAsync(cancellationToken)
+                        .ContinueWith((_) => WriteInternalCoreAsync(array, offset, numBytes, cancellationToken));
```

Comment:
Chaining like this is the right idea.  However, using ContinueWith as is done is going to be problematic.  ContinueWith executes the continuation regardless of whether the antecedent task completes successfully, due to cancellation, or due to faulting with an exception, and we'd only want to do the subsequent write if the flush completed successfully.  If it didn't, we want to return the status from the FlushWriteAsync task instead of that of executing the write.  The semantics of await are more appropriate here.  As such, you probably want a helper async method that awaits the flush and then awaits the write, and that helper can be invoked here.